### PR TITLE
Allow passing CuPy arrays to Matrix replace_coefficients

### PR DESCRIPTION
**Pull request by Yuriyzabegaev**
I've implemented this feature and unit test for it in the same manner as for Matrix upload_CSR method.

### CODE REVIEW

File: pyamgx/Matrix.pyx [hunk 1]
```diff
@@ -182,12 +182,23 @@ cdef class Matrix:
...
+        
+        if hasattr(data, "__array_interface__"):
+            desc = data.__array_interface__
+        elif hasattr(data, "__cuda_array_interface__"):
+            desc = data.__cuda_array_interface__
+        else:
+            raise TypeError("Must pass array-like for data")
+            
+        cdef uintptr_t ptr = desc["data"][0]
+
```
Comment (shwina) [inline pyamgx/Matrix.pyx:194]: Could we use the utility function [`ptr_from_array_interface`](https://github.com/shwina/pyamgx/blob/main/pyamgx/utils.pyx#L5) here instead?

Comment (Yuriyzabegaev) [reply pyamgx/Matrix.pyx:194]: Ok, should I check for dtype here?

Comment (shwina) [reply pyamgx/Matrix.pyx:194]: Yes -- let's do that

**Comment by shwina** [comment]
Awesome! Thanks for your contribution, @Yuriyzabegaev!

**Pull request merged**